File: Cosmos-IOS/Cosmostation/Cosmostation/Model/KavaModel/MyCdp.swift
//
//  MyCdp.swift
//  Cosmostation
//
//  Created by 정용주 on 2021/03/04.
//  Copyright © 2021 wannabit. All rights reserved.
//

import Foundation

public struct MyCdp {
    var cdp: Cdp?
    var collateral_value: Coin?
    var collateralization_ratio: String?
    
    init(_ dictionary: NSDictionary?) {
        if let rawCdp = dictionary?["cdp"] as? NSDictionary {
            self.cdp = Cdp.init(rawCdp)
        }
        if let rawCollateralValue = dictionary?["collateral_value"] as? NSDictionary {
            self.collateral_value = Coin.init(rawCollateralValue)
        }
        self.collateralization_ratio = dictionary?["collateralization_ratio"] as? String
    }
    
    public func getTotalCollateralAmount() -> NSDecimalNumber {
        return NSDecimalNumber.init(string: cdp!.collateral!.amount)
    }
    
    public func getDpCollateralValue(_ pDenom:String) -> NSDecimalNumber {
        return NSDecimalNumber.init(string: collateral_value!.amount).multiplying(byPowerOf10: -WUtils.getKavaCoinDecimal(pDenom))
    }
    
    public func getLiquidationPrice(_ cDenom:String, _ pDenom:String, _ cParam:CollateralParam) -> NSDecimalNumber {
        let collateralAmount = cdp!.getRawCollateralAmount().multiplying(byPowerOf10: -WUtils.getKavaCoinDecimal(cDenom))
        let rawDebtAmount = cdp!.getEstimatedTotalDebt(cParam).multiplying(by: NSDecimalNumber.init(string: cParam.liquidation_ratio)).multiplying(byPowerOf10: -WUtils.getKavaCoinDecimal(pDenom))
        return rawDebtAmount.dividing(by: collateralAmount, withBehavior: WUtils.getDivideHandler(WUtils.getKavaCoinDecimal(pDenom)))
    }
    
    public func getDpEstimatedTotalDebtValue(_ pDenom:String, _ cParam:CollateralParam) -> NSDecimalNumber {
        return cdp!.getEstimatedTotalDebt(cParam).multiplying(byPowerOf10: -WUtils.getKavaCoinDecimal(pDenom))
    }
    
    public func getWithdrawableAmount(_ cDenom:String, _ pDenom:String, _ cParam:CollateralParam, _ cPrice:NSDecimalNumber, _ selfDepositAmount: NSDecimalNumber) -> NSDecimalNumber {
        let cValue = NSDecimalNumber.init(string: collateral_value!.amount)
        let minCValue = cdp!.getEstimatedTotalDebt(cParam).multiplying(by: cParam.getLiquidationRatio()).dividing(by: NSDecimalNumber.init(string: "0.95"), withBehavior:WUtils.handler0Down)
        let maxWithdrawableValue = cValue.subtracting(minCValue)
//            print("maxWithdrawableValue " , maxWithdrawableValue)
        var maxWithdrawableAmount = maxWithdrawableValue.multiplying(byPowerOf10: WUtils.getKavaCoinDecimal(cDenom) - WUtils.getKavaCoinDecimal(pDenom)).dividing(by: cPrice, withBehavior: WUtils.handler0Down)
//            print("maxWithdrawableAmount " , maxWithdrawableAmount)
        
        if (maxWithdrawableAmount.compare(selfDepositAmount).rawValue > 0) {
            maxWithdrawableAmount = selfDepositAmount
        }
        if (maxWithdrawableAmount.compare(NSDecimalNumber.zero).rawValue <= 0) {
            maxWithdrawableAmount = NSDecimalNumber.zero
        }
        return maxWithdrawableAmount
        
        
    }
    
    public func getMoreLoanableAmount(_ cParam:CollateralParam) -> NSDecimalNumber {
        var maxDebtValue = NSDecimalNumber.init(string: collateral_value!.amount).dividing(by: cParam.getLiquidationRatio(), withBehavior: WUtils.handler0Down)
//            print("maxDebtValue " , maxDebtValue)
        maxDebtValue = maxDebtValue.multiplying(by: NSDecimalNumber.init(string: "0.95"), withBehavior: WUtils.handler0Down)
        maxDebtValue = maxDebtValue.subtracting(cdp!.getEstimatedTotalDebt(cParam))
        if (maxDebtValue.compare(NSDecimalNumber.zero).rawValue <= 0) {
            maxDebtValue = NSDecimalNumber.zero
        }
        return maxDebtValue
    }
    
}

File: app/src/main/java/wannabit/io/cosmostaion/model/kava/MyCdp.java
package wannabit.io.cosmostaion.model.kava;

import android.content.Context;

import com.google.gson.annotations.SerializedName;

import java.math.BigDecimal;
import java.math.RoundingMode;

import wannabit.io.cosmostaion.base.BaseData;
import wannabit.io.cosmostaion.model.type.Coin;
import wannabit.io.cosmostaion.utils.WDp;
import wannabit.io.cosmostaion.utils.WUtil;

public class MyCdp {
    @SerializedName("cdp")
    public Cdp cdp;

    @SerializedName("collateral_value")
    public Coin collateral_value;

    @SerializedName("collateralization_ratio")
    public String collateralization_ratio;

    public String getDenom() {
        return cdp.collateral.denom;
    }

    public String getPDenom() {
        return cdp.principal.denom;
    }

    public String getMarketId() {
        return cdp.collateral.denom + ":usd:30";
    }

    public String getDpMarketId() {
        return cdp.collateral.denom.toUpperCase() + " : usdx".toUpperCase() ;
    }

    public String getImagePath() {
        return cdp.collateral.denom + "usd.png";
    }

    public BigDecimal getCollateralAmount() {
        try {
            return new BigDecimal(cdp.collateral.amount);
        } catch (Exception e) {}
        return BigDecimal.ZERO;
    }

    public BigDecimal getPrincipalAmount() {
        try {
            return new BigDecimal(cdp.principal.amount);
        } catch (Exception e) { }
        return BigDecimal.ZERO;
    }

    public BigDecimal getAccumulatedFees() {
        try {
            return new BigDecimal(cdp.accumulated_fees.amount);
        } catch (Exception e) { }
        return BigDecimal.ZERO;
    }

    public BigDecimal getRawDebtAmount() {
        return getPrincipalAmount().add(getAccumulatedFees());
    }

    public BigDecimal getEstimatedTotalFee(Context c, CollateralParam cParam) {
        BigDecimal hiddenFeeValue = WDp.getCdpHiddenFee(c, getRawDebtAmount(), cParam, cdp);
        return  getAccumulatedFees().add(hiddenFeeValue);
    }

    public BigDecimal getEstimatedTotalDebt(Context c, CollateralParam cParam) {
        BigDecimal hiddenFeeValue = WDp.getCdpHiddenFee(c, getRawDebtAmount(), cParam, cdp);
        return  getRawDebtAmount().add(hiddenFeeValue);
    }

    public BigDecimal getLiquidationPrice(Context c, BaseData baseData, CollateralParam cParam) {
        int denomCDecimal = WUtil.getKavaCoinDecimal(baseData, getDenom());
        int denomPDecimal = WUtil.getKavaCoinDecimal(baseData, getPDenom());
        BigDecimal collateralAmount = getCollateralAmount().movePointLeft(denomCDecimal);
        BigDecimal estimatedDebtAmount = getEstimatedTotalDebt(c, cParam).multiply(new BigDecimal(cParam.liquidation_ratio)).movePointLeft(denomPDecimal);
        return estimatedDebtAmount.divide(collateralAmount, denomPDecimal, BigDecimal.ROUND_DOWN);
    }

    public BigDecimal getWithdrawableAmount(Context c, BaseData baseData, CollateralParam cParam, BigDecimal price, BigDecimal selfDeposit) {
        int denomCDecimal = WUtil.getKavaCoinDecimal(baseData, getDenom());
        int denomPDecimal = WUtil.getKavaCoinDecimal(baseData, getPDenom());
        BigDecimal cValue = new BigDecimal(collateral_value.amount);
        BigDecimal minCValue = getEstimatedTotalDebt(c, cParam).multiply(new BigDecimal(cParam.liquidation_ratio)).divide(new BigDecimal("0.95"), 0, BigDecimal.ROUND_DOWN);
        BigDecimal maxWithdrawableValue = cValue.subtract(minCValue);
//            WLog.w("maxWithdrawableValue " + maxWithdrawableValue);
        BigDecimal maxWithdrawableAmount = maxWithdrawableValue.movePointLeft(denomPDecimal - denomCDecimal).divide(price, 0, RoundingMode.DOWN);
//            WLog.w("maxWithdrawableAmount " + maxWithdrawableAmount);

        if (maxWithdrawableAmount.compareTo(selfDeposit) > 0 ) {
            maxWithdrawableAmount =  selfDeposit;
        }
        if (maxWithdrawableAmount.compareTo(BigDecimal.ZERO) <= 0) {
            maxWithdrawableAmount = BigDecimal.ZERO;
        }
        return maxWithdrawableAmount;

    }

    public BigDecimal getMoreLoanableAmount(Context c, CollateralParam cParam) {
        BigDecimal maxDebtValue = new BigDecimal(collateral_value.amount).divide(new BigDecimal(cParam.liquidation_ratio), 0, RoundingMode.DOWN);
//            WLog.w("maxDebtValue " + maxDebtValue);
        maxDebtValue = maxDebtValue.multiply(new BigDecimal("0.95")).setScale(0, RoundingMode.DOWN);
//            WLog.w("maxDebtValue padding " + maxDebtValue);

        maxDebtValue = maxDebtValue.subtract(getEstimatedTotalDebt(c, cParam));
        if (maxDebtValue.compareTo(BigDecimal.ZERO) <= 0) {
            maxDebtValue = BigDecimal.ZERO;
        }
        return maxDebtValue;

    }
}
